### DISCUSSION TITLE
How to cross-compiling wasm32-unknown-unknown on ubuntu

### DISCUSSION THREAD

**Discussion by SoyEstrella** [PyO3/pyo3#2306] [Questions] [answered]
### Bug Description

I imported the pyo3 library into the substrate code(substrate is a framework for building blockchains), however the code cannot be compiled, the error reports under Windows system are as follows:
```
error: failed to run custom build command for `pyo3-ffi v0.16.2`

  Caused by:
    process didn't exit successfully: `C:\Users\19735\Desktop\offchain_worker_2\target\debug\wbuild\node-template-runtime\target\release\build\pyo3-ffi-ee5b2633fd18a1d7\build-script-build` (exit code: 101)
    --- stdout
    cargo:rerun-if-env-changed=PYO3_CROSS
    cargo:rerun-if-env-changed=PYO3_CROSS_LIB_DIR

    --- stderr
    thread 'main' panicked at 'Path does not exist: Os { code: 267, kind: NotADirectory, message: "目录名称无效。
" }', C:\Users\19735\.cargo\registry\src\mirrors.sjtug.sjtu.edu.cn-7a04d2510079875b\pyo3-build-config-0.16.2\src\" }', C:\Users\19impl_.rs:992:33
    note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
  warning: build failed, waiting for other jobs to finish...
  error: build failed
```
So I searched for the solution and I found this issue:[https://github.com/PyO3/pyo3/issues/1221](url)
I switched the operating system to Linux, and followed the instructions of this issue, however I met another error:
`error: your Rust architecture (32-bit) does not match your python interpreter (64-bit)`

I guess the answer is as @davidhewitt said in issue: I should disable check_target_architecture when cross-compiling. But I do not know how to achieve it.

### Steps to Reproduce

1. Compiled Python 3.10.4 with
```
./configure --enable-shared
make
```
2. Setup PYO3_CROSS flags
```
export PYO3_CROSS_PYTHON_VERSION=3.10
export PYO3_CROSS_INCLUDE_DIR=/home/Python-3.10.4/Include
export PYO3_CROSS_LIB_DIR=/home/Python-3.10.4/build/lib.linux-x86_64-3.10
```
3. Part of lib.rs
```
#![cfg_attr(not(feature = "std"), no_std)]

use pyo3::prelude::*;
use pyo3::types::IntoPyDict;

impl<T: Config> Pallet<T> {
	fn hello() -> PyResult<()> {
		Python::with_gil(|py| {
			let sys = py.import("sys")?;
			let version: &str = sys.getattr("version")?.extract()?;
			let locals = [("os", py.import("os")?)].into_py_dict(py);
			let code = "os.getenv('USER') or os.getenv('USERNAME') or 'Unknown'";
			let user: &str = py.eval(code, None, Some(&locals))?.extract()?;
			// println!("Hello {}, I'm Python {}", user, version);
			log::info!("Hello {}, I'm Python {}", user, version);
			Ok(())
		})
	}
}
```

4. Part of Cargo.toml
```
[dependencies]
pyo3 = { version = "0.16.2", default-features = false, features = ["auto-initialize"] }
```

5. build.rs（Wasm setting of substrate）
```
use substrate_wasm_builder::WasmBuilder;

fn main() {
	WasmBuilder::new()
		.with_current_project()
		.export_heap_base()
		.import_memory()
		.build()
}
```

### Your operating system and version

linux ubuntu 20.04.3

### Your Python version (`python --version`)

Python 3.10.4

### Your Rust version (`rustc --version`)

rustc 1.59.0-nightly(6db0a0e9a 2021-11-29)

### Your PyO3 version

0.16.2

### How did you install python? Did you use a virtualenv?

I installed it from the Python official website, and I used vmware virtualenv.

**Comment by davidhewitt** [comment, marked as answer]
Please see https://github.com/PyO3/pyo3/issues/1522#issuecomment-808956769 - you need make sure you are targeting wasm in your Rust build and also have built your own wasm Python interpreter (I'm not aware of any wasm distributions of Python). It sounds like Python 3.11 is the first Python release that might support building for wasm.

Certainly this is something you're free to try but is out of scope for PyO3 for now.
